### DISCUSSION TITLE
Create an FAQ using questions from the github issues

### DISCUSSION THREAD

**Discussion by sm-Fifteen** [fastapi/fastapi#9210] [Questions] [answered]
### Is your feature request related to a problem

A lot of very good questions have been [asked in the FastAPI github issues](https://github.com/tiangolo/fastapi/issues?utf8=%E2%9C%93&q=is%3Aissue+label%3Aquestion+) and some very good answers provided to them, and I worry that those might be harder to find than similar questions on StackOverflow, [where the FastAPI tag hasn't really taken off yet](https://stackoverflow.com/questions/tagged/fastapi).

I've been combing through old notifications using the new Github notification inbox thing, and was pretty impressed at the depth you'd get from some answers, so I figure it might be a good idea to move some of them in the documentation where they might be easier to find.

### The solution you would like

Add an FAQ section to the documentation with a number of quality answers from the github issues.

### Describe alternatives you've considered

I guess we could always reccomend that people start asking questions on Stack Overflow instead? That would probably have the same effect of increasing the "googleability" of those questions and answers, though existing questions wouldn't get to be "migrated".

### Additional context

Call me vain, but I'm pretty proud of [this reply](https://github.com/tiangolo/fastapi/issues/732#issuecomment-558780895) regarding the use and misuse of async in FastAPI/Starlette, [this one](https://github.com/tiangolo/fastapi/issues/726#issuecomment-557687526) about the lifetime dependency pattern (which will hopefully become a standard feature eventually 🤞) and [this one](https://github.com/tiangolo/fastapi/issues/678#issuecomment-549476125) about attaching middleware as parameters instead of decorators (though that last one should probably just be part of the tutorial). I would add other quality answers from other users in there, but while those in my inbox are easy for me to look up and find, there are heaps of good ones I remember and can't find anymore. :/

There are a lot of other questions I've seen in the last couple of months that are very well written and/or in-depth, but may not be a good fit for the (already pretty content-packed) tutorial, and giving them some extra visibility would probably be a good thing. I wouldn't even have found my own questions and answers if it wasn't for the notification inbox, let alone those written by other people.

**Comment by tiangolo** [comment]
I think this is a very relevant idea/issue. :+1: 

On one side, I don't want to add docs for every corner case including some more complex or very advanced use cases that not everyone might need, of course, some extra docs for several cases are needed, but not for all the things that people ask.

---

For example, I think it should be in the docs how to call sync code from async functions using `run_in_threadpool`. That's kind of advanced, but generic enough to apply to many.

Examples of things I think shouldn't go in the docs are things like: "How to deploy at X provider using the specific database Y". Or "how to use FastAPI as if it was the other framework X". Or "how to use the old and strange library X with FastAPI".

The main reasons to not include those examples are that some might be too specific to some use cases, and others might be "strange" use cases that I would not want to encourage by showing them in the docs.

---

On the other side, some of these questions and answers (sometimes advanced) might end up too hidden just in a comment in a closed issue while actually being quite important and relevant for many people. And at the same time, the help many of you are providing here is not necessarily having all the credit it deserves.

I was thinking that maybe trying to move questions to StackOverflow (and promoting answering questions there, for example, [enabling notifications for the tag](https://meta.stackoverflow.com/questions/254318/how-to-subscribe-to-tags)) could be a good idea as good answers get promoted and earn reputation for the people helping there (you).

But at the same time, I'm not sure if it would be more cumbersome to go to StackOverflow and answer questions there instead of keeping them here in GitHub issues.

Asking for opinions of several community members that have been helping a lot here, @dmontagu , @euri10 , @prostomarkeloff , @phy25, @acnebs, what do you think?

Would you prefer to help at StackOverflow than here?

**Comment by euri10** [comment]
all-in for moving questions to stackoverflow !


On Fri, Apr 10, 2020 at 6:55 PM Sebastián Ramírez <notifications@github.com>
wrote:

> I think this is a very relevant idea/issue. 👍
>
> On one side, I don't want to add docs for every corner case including some
> more complex or very advanced use cases that not everyone might need, of
> course, some extra docs for several cases are needed, but not for all the
> things that people ask.
> ------------------------------
>
> For example, I think it should be in the docs how to call sync code from
> async functions using run_in_threadpool. That's kind of advanced, but
> generic enough to apply to many.
>
> Examples of things I think shouldn't go in the docs are things like: "How
> to deploy at X provider using the specific database Y". Or "how to use
> FastAPI as if it was the other framework X". Or "how to use the old and
> strange library X with FastAPI".
>
> The main reasons to not include those examples are that some might be too
> specific to some use cases, and others might be "strange" use cases that I
> would not want to encourage by showing them in the docs.
> ------------------------------
>
> On the other side, some of these questions and answers (sometimes
> advanced) might end up too hidden just in a comment in a closed issue while
> actually being quite important and relevant for many people. And at the
> same time, the help many of you are providing here is not necessarily
> having all the credit it deserves.
>
> I was thinking that maybe trying to move questions to StackOverflow (and
> promoting answering questions there) could be a good idea as good answers
> get promoted and earn reputation for the people helping there (you).
>
> But at the same time, I'm not sure if it would be more cumbersome to go to
> StackOverflow and answer questions there instead of keeping them here in
> GitHub issues.
>
> Asking for opinions of several community members that have been helping a
> lot here, @dmontagu <https://github.com/dmontagu> , @euri10
> <https://github.com/euri10> , @prostomarkeloff
> <https://github.com/prostomarkeloff> , @phy25 <https://github.com/phy25>,
> @acnebs <https://github.com/acnebs>, what do you think?
>
> Would you prefer to help at StackOverflow than here?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/tiangolo/fastapi/issues/992#issuecomment-612119237>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAINSPW5O7FXOPCVZZHSDN3RL5FOTANCNFSM4KWVRI7A>
> .
>


-- 
benoit barthelet
http://pgp.mit.edu/pks/lookup?op=get&search=0xF150E01A72F6D2EE

**Comment by connebs** [comment]
On the one hand, I do like how questions/answers are a bit easier to surface on SO than in GH Issues, and the community / reach is obviously quite a bit larger. On the other hand, I **really** like how the conversations in GH Issues tend to be much more "holistic" than on SO.

On SO, you end up with a bunch of people asking very specific questions, generally to be answered by very specific code which oftentimes does exactly what the asker is looking for, but frequently does not end up actually increasing anyone's understanding. The best SO answers of course also give a very thorough explanation of what is happening, but I think we can all agree that this is often ignored, and people instead just end up copy-pasting code and only dig deeper if that code doesn't immediately work. I also see a lot of posts on SO that don't age well, as the "accepted" answer (with many upvotes) can be accurate at the time the question was asked, but 7 years later when things have changed (for example, let's say people are using Python 3.8 instead of Python 2.7), there will now be a newer, better answer that people often miss because it is not at the top and does not have a green checkmark next to it.

With GH Issues, the fact that it is a thread means that the most up-to-date information will be at the bottom and you can (almost) always rely on that being the case. The ability to very quickly link to relevant source files / pull requests / commits / etc on the same site is also super helpful, again for providing that "holistic" conversation that I think is important for all of us. And ever since GH added emoji responses to comments, Issues threads also work fairly well for the "SO" crowd, who are just looking for a simple code solution to their problem, as the comments with such a copy-paste solution tend to be decorated with lots of ❤️, 👍, and 🎉.

I guess what I really want is for GH to add a QA section that is separate from Issues, so that there can be questions in one dedicated place and bug-reports / feature requests in another. I know it can sometimes be demotivating when you have hundreds of open issues on a project and it doesn't help that many of them are just asking basic questions that are already answered in the docs. Suffice to say I am of two minds on the issue, but generally I prefer the conversations in GH issues to SO these days.

Re: an actual FAQ – one option for more in-depth stuff would be to use GH's wiki feature.

**Comment by tiangolo** [comment]
Ha! Those are very good points @acnebs !

Another :heavy_plus_sign: for GH is that I can actually make sure to sweep all the open issues from time to time and answer anything that's still missing an answer, even if they are oldish.

And another :heavy_plus_sign: is that I can use https://github.com/tiangolo/issue-manager here.

---

I think the best and simplest option, for now, is probably to treat this in a case by case basis, and "promote" common or important questions and answers to docs. Even creating new issues to request including some info from other issues in the docs if necessary.

**Comment by tiangolo** [comment, marked as answer]
I recently moved questions to GitHub Discussions :rocket: That way we get the best of all worlds, I can keep computing FastAPI Experts, and the information is in its rightful place. :tada: 

In fact, I'm just gonna move this one over there too. :nerd_face: 

> Sorry for the long delay! 🙈  I wanted to personally address each issue/PR and they piled up through time, but now I'm checking each one in order.
